[WIP] Prototype for Collaboration

hkirat opened the pull request.

hkirat commented:
Steps to Run
 - npm run start
 - Go to src/extensions/default/bramble/lib/RemoteCommandHandler.js 
 - Set line number 200 ([this line](https://github.com/mozilla/brackets/pull/770/files#diff-ff6f9f7c4c8b8efa6a41925f0fbe77d3R200)) to 
```js
           var delta = args[0];
```
This needs to be done to avoid the problem of same events being relayed again and again causing an infinite loop
 - Open localhost:8000/src/hostel.html
 - Click on Togetherjs button on top (only once, it takes some time to load)
 - Once you get the link to share, DONT OPEN.
 - First go to src/extensions/default/bramble/lib/RemoteCommandHandler.js line 200 and revert it to
```js
           var delta = args[0][0];
```
 - Now open the shared link.
 - Change the contents of the FIRST window. Collaboration should work.

hkirat commented:
Another way of testing is not following the steps above. Simply opening, starting togetherjs and sharing.
If you try this you'll notice that right now events are being re-transmitted again and again.

Eg: User 1 adds a line, User 2 receives this change, User 2 emits this change back to user 1 and so on..
Hence as soon as you do any change, the events transmit in an infinite loop.

humphd commented:
Can you rebase this so the new commits don't show up as changes?

humphd commented on src/hosted.html:
```diff
@@ -21,7 +21,14 @@
-</body>
+    <script>
+        TogetherJS.hub.on('data', function (delta) {
+            console.log("bhai ka aaya data" + delta);
```
Love this, what does it mean?

hkirat replied:
Yup sorry about that. Completely missed that out. 
It means "Here's your data Brother". 

Will get rid of it.

humphd commented on src/extensions/default/bramble/lib/RemoteCommandHandler.js:
```diff
@@ -194,6 +194,30 @@ define(function (require, exports, module) {
             CommandManager.execute("bramble.addCodeSnippet", args[0]).always(callback);
             break;
+        case "CODEMIRROR_RELAY":
+            var editor = EditorManager.getActiveEditor();
```
We need to move all this into the editor, probably `Editor.js`, vs. doing it across the bramble `postMessage` API.

gideonthomas replied:
+1 to this. Have a separate file in the same folder as `Editor.js` (maybe `Collaboration.js`) and put all this code in there. Then just require it in `Editor.js` and call a function to initialize it or something like that.

hkirat commented:
Fixed the loop problem @gideonthomas. 
Working good now. The Auto inundation addon is creating a few issues, but if you turn it off it works great.

gideonthomas commented on src/editor/Editor.js:
```diff
@@ -885,6 +887,18 @@ define(function (require, exports, module) {
         this._updateHiddenLines();
     };
 
+    Editor.prototype.mutualExcluse = function (f, changeList) {
```
what does all of this do? Please add a comment explaining.

fpbird replied:
It makes sure that if a togetherjs event has triggered a change in the codemirror instance of this particular editor, then it shouldn't transmit a new event to all the connected clients.
I saw this in y-js's collaboration code. 
But I think it should not be implemented as a prototype function to the editor, more like a global function.

hkirat replied:
It makes sure that if a togetherjs event has triggered a change in the codemirror instance of this particular editor, then it shouldn't transmit a new event to all the connected clients.
I saw this in y-js's collaboration code.
But I think it should not be implemented as a prototype function to the editor, more like a global function.

flukeout commented:
Hey @hkirat - is this ready to test and experiment with? I'll spend some time today testing it out.

gideonthomas commented:
@flukeout we're still working on getting the code in the right shape.

hkirat commented:
Yes @flukeout as gideon mentioned the code doesn't checkout so well. 
Though it works okay. You can play around and try if you want.

gideonthomas commented on src/editor/Editor.js:
```diff
@@ -77,7 +77,8 @@ define(function (require, exports, module) {
         MainViewManager    = require("view/MainViewManager"),
-        _                  = require("thirdparty/lodash");
+        _                  = require("thirdparty/lodash"),
+        brambleEvents	   = require("bramble/BrambleEvents");
```
In Collaboration.js:

```js
function Collaboration() {
   ...
}

Collaboration.prototype.create = function() {}

...

exports.Collaboration = Collaboration;
```

gideonthomas replied:
In Editor.js:

```js
var Collaboration = require("editor/Collaboration");

function Editor() {
    ...
    this._codemirror = ...;
    this.collaboration = new Collaboration(this._codemirror);
}
```

gideonthomas replied:
In Editor.js:
```js
function Editor(..., collaboration) {
   this._codemirror = ...;
   collaboration.setCodemirror(this._codemirror);
```

gideonthomas commented on src/editor/Editor.js:
```diff
@@ -928,6 +942,7 @@ define(function (require, exports, module) {
         // whereas the "change" event should be listened to on the document. Also the
         // Editor dispatches a change event before this event is dispatched, because
         // CodeHintManager needs to hook in here when other things are already done.
+        this.mutualExcluse(brambleEvents.triggerCodeMirrorChange, changeList);
```
```js
collabinstance.triggerCodemirrorchange(changeList);
```

hkirat commented:
Hey @gideonthomas 
I was working on the things we discussed yesterday. requirejs doesn't seem to load togetherjs as expected.
I dont think it loads caman as well. 
If you observe [this](https://github.com/mozilla/brackets/blob/master/src/editor/Image.js#L4) during execution, the value of caman remains undefined.

gideonthomas commented:
@hkirat what's the status of this?

hkirat commented:
After modifying the code as we discussed, I ran into the following problems: 
1. The link as generated by togetherjs is on the path /src/index.html. This might be because we are calling Togetherjs() from inside a js file that is loaded in index.html
I tried looking into the code of togetherjs. There is a line that uses the Location variable to determine the route of the page, however modifying this to "/src/hosted.html" made togetherjs break.
2. Togetherjs doesn't function as expected. Even in one window I see a mouse pointer copying the motion of my mouse pointer.

I think loading together.min.js at the top level director (hosted.html/hosted.js) will fix this. However I think that is against the practices followed in brackets. 
Even if we do load together.min.js in hosted.html/hosted.js, the togetherjs variable will not be accessible in Collaboration.js. 
Any ideas on how we should proceed?

gideonthomas commented:
1. The link that it generates for `src/index.html` is correct, it should be for that path (for now). We can change it to be configurable later. Until then, this obviously creates a problem in terms of how to access the togetherjs link from another client. you might want to edit the hosted.js file so that we allow this - i.e. accept something via the url (like the generated togetherjs link) to use instead of `src/index.html`
2. I don't understand what you're seeing. Could you elaborate some more?

hkirat commented:
As per the implementation right now, on opening brackets something like this happens: 

![vid](https://media.giphy.com/media/bvJDbAianH8ze/giphy.gif)

Togetherjs seems to interpret my pointer as a foreign pointer.

I see some logs along: 
![screenshot from 2017-06-05 11 41 50](https://cloud.githubusercontent.com/assets/8079861/26772579/0045bdd6-49e4-11e7-8d57-b431a173ed03.png)

hkirat commented:
These are the events that happen on each click, one Send and two In events
![screenshot from 2017-06-05 15 26 08](https://cloud.githubusercontent.com/assets/8079861/26779553/6d921a78-4a03-11e7-8bbd-ecf0c89cf268.png)
It looks like for every event sent from the page, togetherjs receives the same event twice.
This is what happens on jsfiddle on each click: 
![screenshot from 2017-06-05 15 29 02](https://cloud.githubusercontent.com/assets/8079861/26779615/b8bc4d7a-4a03-11e7-97ef-6fb17835d32a.png)

humphd commented on src/editor/Collaboration.js:
```diff
@@ -0,0 +1,56 @@
+             return;
+            }
+            self.changing = true;
+            var cm = this.codeMirror;
```
`this.codeMirror` should be `self.codeMirror`

hkirat replied:
Ah finally.
Makes sense why it wasn't working on chrome and was working on Safari.
Thanks for pointing this out.

humphd commented on src/editor/Collaboration.js:
```diff
@@ -0,0 +1,56 @@
+    };
+
+    Collaboration.prototype.setCodemirror = function(codemirror) {
+       this.codemirror = codemirror;
```
Move this into `init` so you don't have to the check for `self.codemirror` on 19 above.

hkirat replied:
Had to remove this function completely for now. Apparently togetherjs handles syncing of codemirror instances. [Reference](https://github.com/mozilla/togetherjs/blob/develop/togetherjs/forms.js#L180)
So all the code in Collaboration.js became redundant.

hkirat commented on src/editor/Collaboration.js:
```diff
@@ -0,0 +1,15 @@
+    function Collaboration() {
+        this.togetherjs = togetherjs;
+        var self = this;
+        window.setTimeout(function() {
```
Added this over here to make sure that togetherjs loads after brackets has loaded completely. 
After getting loaded, togetherjs looks for all the codemirror instances to keep in sync.
Will probably move to something like
```js
window.onload
```
or
```js
document.onload
```
later.

humphd replied:
Using a timeout is a bad idea, let's just use existing events, or add a new one.  What are you trying to achieve exactly?

* There are a bunch of handlers for various startup events here: https://github.com/mozilla/brackets/blob/master/src/utils/AppInit.js#L130-L147
* https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteEvents.js#L237-L273 is when we are 100% loaded and running as a usable editor, and we could easily emit an internal event here

hkirat commented:
@gideonthomas As we discussed, togetherjs is *almost* similar to the one present on the website.
To include it via requirejs, I've added a 
```js
define
```
in the beginning and removed the function call in the end and returned TogetherJS from the module.
See [this](https://github.com/hkirat/brackets/blob/6eff29c9cd9b7d2b40cecb80c750f133e034a0b9/src/thirdparty/togetherjs/together.js#L6) and [this](https://github.com/hkirat/brackets/blob/6eff29c9cd9b7d2b40cecb80c750f133e034a0b9/src/thirdparty/togetherjs/together.js#L792)
Is there some other way to load Togetherjs via require using the code directly present [here](https://togetherjs.com/togetherjs-min.js)

hkirat commented:
Closing this as we'll be shifting from togetherjs to other libraries as discussed with gideon.
